server/src/main/java/org/elasticsearch/search/aggregations/pipeline/PercentileRanksBucketPipelineAggregator.java

```diff
@@ -0,0 +1,101 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.pipeline;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.pipeline.BucketHelpers.GapPolicy;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+public class PercentileRanksBucketPipelineAggregator extends BucketMetricsPipelineAggregator {
+
+    private final double[] percents;
+    private boolean keyed = true;
+    private List<Double> data;
+
+    PercentileRanksBucketPipelineAggregator(String name, double[] percents, boolean keyed, String[] bucketsPaths,
+                                        GapPolicy gapPolicy, DocValueFormat formatter, Map<String, Object> metaData) {
+        super(name, bucketsPaths, gapPolicy, formatter, metaData);
+        this.percents = percents;
+        this.keyed = keyed;
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public PercentileRanksBucketPipelineAggregator(StreamInput in) throws IOException {
+        super(in);
+        percents = in.readDoubleArray();
+        keyed = in.readBoolean();
+    }
+
+    @Override
+    public void innerWriteTo(StreamOutput out) throws IOException {
+        out.writeDoubleArray(percents);
+        out.writeBoolean(keyed);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return PercentileRanksBucketPipelineAggregationBuilder.NAME;
+    }
+
+    @Override
+    protected void preCollection() {
+       data = new ArrayList<>(1024);
+    }
+
+    @Override
+    protected void collectBucketValue(String bucketKey, Double bucketValue) {
+        data.add(bucketValue);
+    }
+
+    @Override
+    protected InternalAggregation buildAggregation(List<PipelineAggregator> pipelineAggregators, Map<String, Object> metadata) {
+
+        // Perform the sorting and percentile collection now that all the data
+        // has been collected.
+        Collections.sort(data);
+        int n = data.size() - 1;
+        double[] PercentileRanks = new double[percents.length];
+        if (data.size() == 0) {
+            for (int i = 0; i < percents.length; i++) {
+                PercentileRanks[i] = Double.NaN;
+            }
+        } else {
+            for (int i = 0; i < percents.length; i++) {
```

Comment:
I don't think this algorithm is correct, or maybe I'm reading it wrong.  Mathematically, percentile ranks is:

` p = k / n`, where:

- `p` is percentile rank
- `k` is number of elements smaller than the value of interest
- `n` is total number of elements

So if we had the sorted array `data: [1,2,4,6,8,10,11,12,14,16,20]` and wanted to know the rank of `values: [13]`,
it would be `p = 8 / 11 = 72.73%`.  

I think we need to:
1. Loop over `values` array
2. for each value, binary search `data` to find the position in the array that the value would sit
3. use that index to calculate the percentile rank

This is a nice little web calculator for percentile ranks if you want to play around: https://mathcracker.com/percentile-rank-calculator.php

Reply:
I think i got a formula and understanding from a different source. will take a look at this.